perception/autoware_tensorrt_bevdet/src/nvjpegdecoder.cpp

```diff
@@ -0,0 +1,299 @@
+      // in the case of rgb create 3 buffers with sizes of original image
+      channels = 3;
+      widths[1] = widths[2] = widths[0];
+      heights[1] = heights[2] = heights[0];
+    }
+
+    if (img_width[i] != widths[0] || img_height[i] != heights[0]) {
+      img_width[i] = widths[0];
+      img_height[i] = heights[0];
+      // realloc output buffer if required
+      for (int c = 0; c < channels; c++) {
+        int aw = mul * widths[c];
+        int ah = heights[c];
+        size_t sz = aw * ah;
+        ibuf[i].pitch[c] = aw;
+        if (sz > isz[i].pitch[c]) {
+          if (ibuf[i].channel[c]) {
+            CHECK_CUDA(cudaFree(ibuf[i].channel[c]));
+          }
+          CHECK_CUDA(cudaMalloc(reinterpret_cast<void **>(&ibuf[i].channel[c]), sz));
+          isz[i].pitch[c] = sz;
+        }
+      }
+    }
+  }
+  return EXIT_SUCCESS;
+}
+
+void create_decoupled_api_handles(std::vector<decode_params_t> & params, share_params & share_param)
+{
+  for (size_t i = 0; i < params.size(); i++) {
+    CHECK_NVJPEG(nvjpegDecoderCreate(
+      share_param.nvjpeg_handle, NVJPEG_BACKEND_DEFAULT, &params[i].nvjpeg_decoder));
+    CHECK_NVJPEG(nvjpegDecoderStateCreate(
+      share_param.nvjpeg_handle, params[i].nvjpeg_decoder, &params[i].nvjpeg_decoupled_state));
+
+    CHECK_NVJPEG(
+      nvjpegBufferPinnedCreate(share_param.nvjpeg_handle, NULL, &params[i].pinned_buffers[0]));
+    CHECK_NVJPEG(
+      nvjpegBufferPinnedCreate(share_param.nvjpeg_handle, NULL, &params[i].pinned_buffers[1]));
+    CHECK_NVJPEG(
+      nvjpegBufferDeviceCreate(share_param.nvjpeg_handle, NULL, &params[i].device_buffer));
+
+    CHECK_NVJPEG(nvjpegJpegStreamCreate(share_param.nvjpeg_handle, &params[i].jpeg_streams[0]));
+    CHECK_NVJPEG(nvjpegJpegStreamCreate(share_param.nvjpeg_handle, &params[i].jpeg_streams[1]));
+
+    CHECK_NVJPEG(
+      nvjpegDecodeParamsCreate(share_param.nvjpeg_handle, &params[i].nvjpeg_decode_params));
+  }
+}
+
+void destroy_decoupled_api_handles(
+  std::vector<decode_params_t> & params, share_params & share_param)
+{
+  for (size_t i = 0; i < params.size(); i++) {
+    CHECK_NVJPEG(nvjpegDecodeParamsDestroy(params[i].nvjpeg_decode_params));
+
+    CHECK_NVJPEG(nvjpegJpegStreamDestroy(params[i].jpeg_streams[0]));
+    CHECK_NVJPEG(nvjpegJpegStreamDestroy(params[i].jpeg_streams[1]));
+    CHECK_NVJPEG(nvjpegBufferPinnedDestroy(params[i].pinned_buffers[0]));
+    CHECK_NVJPEG(nvjpegBufferPinnedDestroy(params[i].pinned_buffers[1]));
+    CHECK_NVJPEG(nvjpegBufferDeviceDestroy(params[i].device_buffer));
+
+    CHECK_NVJPEG(nvjpegJpegStateDestroy(params[i].nvjpeg_decoupled_state));
+    CHECK_NVJPEG(nvjpegDecoderDestroy(params[i].nvjpeg_decoder));
+  }
+}
+
+void release_buffers(std::vector<nvjpegImage_t> & ibuf)
+{
+  for (size_t i = 0; i < ibuf.size(); i++) {
+    for (size_t c = 0; c < NVJPEG_MAX_COMPONENT; c++)
+      if (ibuf[i].channel[c]) CHECK_CUDA(cudaFree(ibuf[i].channel[c]));
+  }
+}
+
+int get_img(
```

Comment:
The `get_img` function performs multiple `cudaMemcpy2D` calls. For improved performance, investigate the possibility of using a single `cudaMemcpy2D` call with appropriate offsets to copy all three channels at once.  This would reduce the overhead of multiple kernel launches.